### PR TITLE
Add a `trainer.ckpt_path` setter for stateful loading

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

```python
trainer = Trainer()
trainer.ckpt_path = "my/checkpoint/file.ckpt"
trainer.fit(model)

trainer.test()  # still uses the `ckpt_path` above

# you will be in charge of resetting this
trainer.ckpt_path = None
trainer.test()  # now uses `ckpt_path="best"` as it would normally
```

This is a tool that NeMo can use in replacement of the `resume_checkpoint_path` attribute, removed in https://github.com/Lightning-AI/lightning/pull/16167: https://github.com/NVIDIA/NeMo/blob/85cc023adc530de59dd13903c4541ca008b260d3/nemo/utils/exp_manager.py#L575

cc @SeanNaren @titu1994

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

File: src/pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 1]
```diff
@@ -391,9 +422,15 @@ def restore_lr_schedulers(self) -> None:
         for config, lrs_state in zip(self.trainer.lr_scheduler_configs, lr_schedulers):
             config.scheduler.load_state_dict(lrs_state)
 
-    # ----------------------------------
-    # PRIVATE OPS
-    # ----------------------------------
+    def _restore_modules_and_callbacks(self, checkpoint_path: Optional[_PATH] = None) -> None:
```
Comment (carmocca) [inline src/pytorch_lightning/trainer/connectors/checkpoint_connector.py:425]: This is just moved off the `trainer.py` file with no changes. Unrelated to the rest of the PR

**Review by Borda — APPROVED** [review]

File: src/pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 2]
```diff
@@ -73,7 +75,7 @@ def resume_start(self, checkpoint_path: Optional[_PATH] = None) -> None:
         3. from `checkpoint_path` file if provided
         4. don't restore
         """
-        self.resume_checkpoint_path = checkpoint_path
```
Comment (awaelchli) [inline src/pytorch_lightning/trainer/connectors/checkpoint_connector.py:76]: Let's keep this. It communicates that this is the checkpoint path to resume from, not to be confused with the one being saved. It gets managed by `resume_start()` and `resume_end()`

Comment (carmocca) [reply src/pytorch_lightning/trainer/connectors/checkpoint_connector.py:76]: This is still there, but renamed to `_ckpt_path`. Are you suggesting to make it public? Or to keep the old name?

I would like that this name matches the one in the trainer, so that would mean having `trainer.resume_checkpoint_path` which then would be inconsistent with `trainer.test(ckpt_path)|

Comment (awaelchli) [reply src/pytorch_lightning/trainer/connectors/checkpoint_connector.py:76]: > Are you suggesting to make it public? Or to keep the old name? 

Both. The existing attribute already sort of serves the same purpose, it is why it was public. We made `resume_checkpoint_path` so that it can be changed if needed.

**Review by awaelchli — APPROVED** [review]
I don't understand these changes. We were supposed to make things simpler, not more complicated.
